File: 2692-take-gifts-from-the-richest-pile/take-gifts-from-the-richest-pile.py
class Solution(object):
    def pickGifts(self, gifts, k):
        gifts = [-gift for gift in gifts]
        heapq.heapify(gifts)

        for _ in range(k):
            max_gift = heapq.heappop(gifts)
            heapq.heappush(gifts, -floor(sqrt(-max_gift)))
        
        return int(-sum(gifts))
        
        

File: LeetCode/2692-take-gifts-from-the-richest-pile/take-gifts-from-the-richest-pile.java
class Solution {
    public long pickGifts(int[] gifts, int k) {
        PriorityQueue<Integer> pQ = new PriorityQueue<>(Collections.reverseOrder());

        for(int gift: gifts){
            pQ.add(gift);
        }

        for(int i = 0; i < k; i++){
            int maxGift = (int)Math.sqrt(pQ.poll());
            pQ.add(maxGift);
        }

        return sumOperation(pQ);
    }

    private long sumOperation(PriorityQueue<Integer> pQ){
        long sum = 0;
        
        for(int num: pQ){
            sum += num;
        }

        return sum;
    }
}